### PR TITLE
Check the modulus bit length

### PR DESCRIPTION
**Pull request by bernd-edlinger**
The check was missing in DH_check and DH_check_params.
DH_check_pub_key_ex was accidentally calling DH_check,
so results were undefined.

[extended tests]

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/dh/dh_lib.c [hunk 1]
```diff
@@ -235,11 +235,11 @@ void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key)
 int DH_set0_key(DH *dh, BIGNUM *pub_key, BIGNUM *priv_key)
 {
     if (pub_key != NULL) {
-        BN_free(dh->pub_key);
+        BN_clear_free(dh->pub_key);
```
Comment (paulidale) [inline crypto/dh/dh_lib.c]: Does `pub_key` really need to be cleared?

Comment (bernd-edlinger) [reply crypto/dh/dh_lib.c]: No, but it is clear free'd in DH_free together with even more public
things like g, p, q, etc. but I would not like to change DH_free to use BN_free.
So this is for consistency.

**Comment by bernd-edlinger** [comment]
Can I cherry-pick the commit
fc0f536 "DH_check_pub_key_ex was accidentally calling DH_check"  to 1.1.1 
and 
fdde60f  "Use BN_clear_free in DH_set0_key"  to all 1.1.1, 1.1.0 and 1.0.2

**Comment by levitte** [comment]
Is this for master only? If not, it should be labeled appropriately

**Comment by bernd-edlinger** [comment]
Yes, sorry for confusion, but whenever I look somewhere, something comes up.

File: crypto/dh/dh_lib.c (hunk 1 above)
Comment (paulidale) [reply crypto/dh/dh_lib.c]: It's okay, I was just querying the change which seemed odd.  I'm happy to clear and free this, it is always safer.

**Comment by paulidale** [comment]
I'm okay with cherry picking the changes.

File: crypto/dh/dh_lib.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/dh/dh_lib.c]: It seems strange to me that DH_free does that for public fields

**Comment by bernd-edlinger** [comment]
Merged to all branches where applicable, cherry-pick for 1.0.2 did fail,
maybe I'll make another PR for that.  Thanks!

File: crypto/dh/dh_lib.c (hunk 1 above)
Comment (bernd-edlinger) [reply crypto/dh/dh_lib.c]: Yes. me too.

**Pull request closed without merging**